### ISSUE TITLE
DynamoDB attribute on a property is not inherited in the object persistence model

### ISSUE THREAD

**Issue by pjh0463** [aws/aws-sdk-net#2590]
### Describe the bug

When using the AWS SDK .NET with DynamoDB, attributes on properties are not inherited in the object persistence model, causing unexpected behavior when serializing and deserializing objects.

### Expected Behavior

DynamoDB attributes on properties should be inherited in the object persistence model, allowing for correct serialization and deserialization of objects.

### Current Behavior

DynamoDB attributes on properties are not inherited in the object persistence model.

### Reproduction Steps

1. Create a parent class with a property decorated with a DynamoDBProperty attribute (or any DynamoDB attribute).
2. Create a child class that inherits from the parent class and overrides the property.
3. The SDK does not recognize the attribute in the property in the inherited class.

### Possible Solution

`Amazon.DynamoDBv2.DataModel.Utils.GetAttributeObjects` should use `CustomAttributeExtensions.GetCustomAttributes` instead of `MemberInfo.GetCustomAttributes`.

`MemberInfo.GetCustomAttributes` ignores the inherit parameter for properties and events. `Attribute.GetCustomAttributes` should be used instead.
https://learn.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.getcustomattributes

`TypeFactory.TypeInfoWrapper` in `AWSSDK.Core` already uses `CustomAttributeExtensions.GetCustomAttributes` so the attribute inheritance works as expected when declared on classes.

Current Code
```c#
private static object[] GetAttributeObjects(MemberInfo targetMemberInfo)
{
    if (targetMemberInfo == null) throw new ArgumentNullException("targetMemberInfo");
#if NETSTANDARD
    object[] attributes = targetMemberInfo.GetCustomAttributes(typeof(DynamoDBAttribute), true).ToArray();
#else
    ...
#endif
    return attributes;
}
``` 

Suggested Code
```c#
private static object[] GetAttributeObjects(MemberInfo targetMemberInfo)
{
    if (targetMemberInfo == null) throw new ArgumentNullException("targetMemberInfo");
#if NETSTANDARD
    object[] attributes = CustomAttributeExtensions.GetCustomAttributes(targetMemberInfo, typeof(DynamoDBAttribute), true).ToArray<object>();
#else
    ...
#endif
    return attributes;
}
```

### AWS .NET SDK and/or Package version used

AWSSDK.DynamoDBv2 3.7.102.31

### Targeted .NET Platform

.Net 6

### Operating System and version

Windows

**Comment by ashishdhingra** [comment]
Hi @pjh0463,

Good afternoon.

Thanks for reporting the issue. Unfortunately, I'm unable to reproduce the issue. Below are the reproduction steps:
- Created DynamoDB table named `DigitalProductCatalog` with `Id` (type `Number`) as partition key.
- Created a console application (reference ) with the following code:
**.csproj**
```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.DynamoDBv2" Version="3.7.102.33" />
  </ItemGroup>

</Project>
```
**Program.cs**
```C#
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using Amazon.Runtime;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace DynamoDBTest
{
    class Program
    {
        private static AmazonDynamoDBClient client = new AmazonDynamoDBClient();

        static void Main(string[] args)
        {
            try
            {
                DynamoDBContext context = new DynamoDBContext(client);
                Task.WaitAll(TestCRUDOperations(context));
                Console.WriteLine("To continue, press Enter");
                Console.ReadLine();
            }
            catch (AmazonDynamoDBException e) { Console.WriteLine(e.Message); }
            catch (AmazonServiceException e) { Console.WriteLine(e.Message); }
            catch (Exception e) { Console.WriteLine(e.Message); }
        }

        private static async Task TestCRUDOperations(DynamoDBContext context)
        {
            int bookID = 1001; // Some unique value.
            DigitalBook myBook = new DigitalBook
            {
                Id = bookID,
                Title = "object persistence-AWS SDK for.NET SDK-Book 1001",
                ISBN = "111-1111111001",
                BookAuthors = new List<string> { "Author 1", "Author 2" },
                ImageUrl = "http://localhost/test.jpg"
            };

            // Save the book.
            await context.SaveAsync(myBook);
            // Retrieve the book.
            DigitalBook bookRetrieved = await context.LoadAsync<DigitalBook>(bookID);

            // Update few properties.
            bookRetrieved.ISBN = "222-2222221001";
            bookRetrieved.BookAuthors = new List<string> { " Author 1", "Author x" }; // Replace existing authors list with this.
            await context.SaveAsync(bookRetrieved);

            // Retrieve the updated book. This time add the optional ConsistentRead parameter using DynamoDBContextConfig object.
            DigitalBook updatedBook = await context.LoadAsync<DigitalBook>(bookID, new DynamoDBContextConfig
            {
                ConsistentRead = true
            });

            // Delete the book.
            await context.DeleteAsync<DigitalBook>(bookID);
            // Try to retrieve deleted book. It should return null.
            Book deletedBook = await context.LoadAsync<DigitalBook>(bookID, new DynamoDBContextConfig
            {
                ConsistentRead = true
            });
            if (deletedBook == null)
                Console.WriteLine("Book is deleted");
        }
    }

    [DynamoDBTable("ProductCatalog")]
    public class Book
    {
        [DynamoDBHashKey] //Partition key
        public int Id
        {
            get; set;
        }
        [DynamoDBProperty]
        public string Title
        {
            get; set;
        }
        [DynamoDBProperty]
        public string ISBN
        {
            get; set;
        }
        [DynamoDBProperty("Authors")] //String Set datatype
        public List<string> BookAuthors
        {
            get; set;
        }
    }

    [DynamoDBTable("DigitalProductCatalog")]
    public class DigitalBook : Book
    {
        // Property override
        [DynamoDBProperty]
        public new string Title
        {
            get;set;
        }

        [DynamoDBProperty]
        public string ImageUrl
        {
            get; set;
        }
    }
}
```

The program executes successfully using the child class.

Please let me know if I'm missing anything. Kindly share the reproduction code and other details in case your scenario is different.

Thanks,
Ashish

**Comment by pjh0463** [comment]
Please see below for the reproduction code. 

```c#
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using Amazon.DynamoDBv2.DocumentModel;

var client = new AmazonDynamoDBClient();
var context = new DynamoDBContext(client);

var parent = new Parent();
parent.Property1 = "Value";

var child = new Child();
child.Property1 = "Value";

// Child DynamoDBPropertyAttribute.AttributeName is not used when converting to Document.
var parentDocument = context.ToDocument(parent);
var childDocument = context.ToDocument(child);

Console.WriteLine($"Parent Document: {GetProperties(parentDocument)}");
Console.WriteLine($"Child Document: {GetProperties(childDocument)}");

// DynamoDBPropertyAttribute is inherited when checked using the correct reflection method.
var childDynamoDbAttributes = Attribute.GetCustomAttributes(
    typeof(Child).GetProperty("Property1"),
    typeof(DynamoDBPropertyAttribute));
var childDynamoDbAttributeNames = string.Join(", ", childDynamoDbAttributes.Select(a => a.GetType().Name));

Console.WriteLine($"Child DynamoDBPropertyAttributes: {childDynamoDbAttributeNames}");

Console.ReadKey();

string GetProperties(Document document)
{
    return string.Join(", ", document.Select(p => $"{p.Key}: {p.Value}"));
}

public class Parent
{
    [DynamoDBProperty("actualPropertyName")]
    public virtual string Property1 { get; set; }
}

public class Child : Parent
{
    public override string Property1 { get; set; }
}
```

**Comment by ashishdhingra** [comment]
@pjh0463 Thanks for your reply. I tried the below code:
```C#
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using Amazon.DynamoDBv2.DocumentModel;
using Amazon.Runtime;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DynamoDBTest
{
    class Program
    {
        private static AmazonDynamoDBClient client = new AmazonDynamoDBClient();

        static void Main(string[] args)
        {
            try
            {
                DynamoDBContext context = new DynamoDBContext(client);

                int bookID = 1001; // Some unique value.
                Book book = new Book
                {
                    Id = bookID,
                    Title = "object persistence-AWS SDK for.NET SDK-Book 1001",
                    ISBN = "111-1111111001",
                    BookAuthors = new List<string> { "Author 1", "Author 2" },
                };

                DigitalBook digitalBook = new DigitalBook
                {
                    Id = bookID,
                    Title = "object persistence-AWS SDK for.NET SDK-Book 1001",
                    ISBN = "111-1111111001",
                    BookAuthors = new List<string> { "Author 1", "Author 2" },
                    ImageUrl = "http://localhost/test.jpg"
                };

                var bookDocument = context.ToDocument(book);
                var digitalBookDocument = context.ToDocument(digitalBook);

                var parentDynamoDbAttributes = Attribute.GetCustomAttributes(typeof(Book).GetProperty("Title"), typeof(DynamoDBPropertyAttribute));
                Console.WriteLine($"Parent DynamoDBPropertyAttributes: {string.Join("\n", parentDynamoDbAttributes.Select(a => string.Join(": ", a.GetType().Name, ((DynamoDBPropertyAttribute)a).AttributeName)))}");

                var childDynamoDbAttributes = Attribute.GetCustomAttributes(typeof(DigitalBook).GetProperty("Title"), typeof(DynamoDBPropertyAttribute));
                Console.WriteLine($"Child DynamoDBPropertyAttributes: {string.Join("\n", childDynamoDbAttributes.Select(a => string.Join(": ", a.GetType().Name, ((DynamoDBPropertyAttribute)a).AttributeName)))}");

                
                Console.WriteLine("~~~~~~~~~~~~~~~~~~~~~~~~~~");
                Console.WriteLine($"**Parent Document**\n{GetProperties(bookDocument)}");
                Console.WriteLine("--------------------------");
                Console.WriteLine($"**Child Document**\n{GetProperties(digitalBookDocument)}");


                Console.WriteLine("To continue, press Enter");
                Console.ReadLine();
            }
            catch (AmazonDynamoDBException e) { Console.WriteLine(e.Message); }
            catch (AmazonServiceException e) { Console.WriteLine(e.Message); }
            catch (Exception e) { Console.WriteLine(e.Message); }
        }

        private static string GetProperties(Document document)
        {
            return string.Join("\n", document.Select(p => $"{p.Key}: {p.Value}"));
        }
    }

    [DynamoDBTable("ProductCatalog")]
    public class Book
    {
        [DynamoDBHashKey] //Partition key
        public int Id
        {
            get; set;
        }
        [DynamoDBProperty("title")]
        public virtual string Title
        {
            get; set;
        }
        [DynamoDBProperty]
        public string ISBN
        {
            get; set;
        }
        [DynamoDBProperty("Authors")] //String Set datatype
        public List<string> BookAuthors
        {
            get; set;
        }
    }

    [DynamoDBTable("DigitalProductCatalog")]
    public class DigitalBook : Book
    {
        [DynamoDBProperty]
        public override string Title
        {
            get; set;
        }

        [DynamoDBProperty]
        public string ImageUrl
        {
            get; set;
        }
    }
}
```

It produced the below output:
```
Parent DynamoDBPropertyAttributes: DynamoDBPropertyAttribute: title
Child DynamoDBPropertyAttributes: DynamoDBPropertyAttribute:
~~~~~~~~~~~~~~~~~~~~~~~~~~
**Parent Document**
Id: 1001
title: object persistence-AWS SDK for.NET SDK-Book 1001
ISBN: 111-1111111001
Authors: Amazon.DynamoDBv2.DocumentModel.PrimitiveList
--------------------------
**Child Document**
Title: object persistence-AWS SDK for.NET SDK-Book 1001
ImageUrl: http://localhost/test.jpg
Id: 1001
ISBN: 111-1111111001
Authors: Amazon.DynamoDBv2.DocumentModel.PrimitiveList
```

If you notice above, even though the child subclass `DigitalBook` inherits the `DynamoDBProperty` attribute information for property `Title`, it doesn't retain the attribute property values (in this case `AttributeName`). I'm unsure even if we make the proposed change, how in such circumstances we would able to construct `Document` property name(s) correctly given that the attribute property `AttributeName` value is lost.

Thanks,
Ashish

**Comment by ashishdhingra** [comment]
As reported by user, per [MemberInfo.GetCustomAttributes Method](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.getcustomattributes?view=net-7.0), `inherit` parameter is ignored for properties and events.

**Comment by ashishdhingra** [comment]
This was manually merged and released today as [version 3.7.103 of AWSSDK.DynamoDBv2](https://www.nuget.org/packages/AWSSDK.DynamoDBv2/3.7.103).

https://github.com/aws/aws-sdk-net/commit/28c5643775c22bf335ca91a0ed03ccc0d0ebafc3
https://github.com/aws/aws-sdk-net/commit/aa8d77d2d47414c3e732da9c161b6ab505a0ceda
